pulsar-zookeeper-utils/src/main/java/org/apache/pulsar/zookeeper/ZkIsolatedBookieEnsemblePlacementPolicy.java

```diff
@@ -147,6 +157,11 @@ private ZooKeeperCache getAndSetZkCache(Configuration conf) {
             Map<String, byte[]> customMetadata, List<BookieId> currentEnsemble,
             BookieId bookieToReplace, Set<BookieId> excludeBookies)
             throws BKNotEnoughBookiesException {
+        // parse the ensemble placement policy from the custom metadata, if it is present, we will apply it to
+        // the isolation groups for filtering the bookies.
+        Optional<EnsemblePlacementPolicyConfig> ensemblePlacementPolicyConfig =
+            getEnsemblePlacementPolicyConfig(customMetadata);
+        ensemblePlacementPolicyConfig.ifPresent(this::fillIsolationGroupWithEnsemblePlacementPolicyConfig);
```

Comment:
The change here has concurrent access issues. Because it resets the primary and secondary isolation groups. It will cause replacing bookies using the wrong group. 

Also, I don't think the logic here is correct.

I think we should record the `group` information in the custom metadata. when replacing a bookie, we need to lookup the bookies in the same group. If it is not an easy change to make here, we should try to build a new bookie ensemble placement policy to achieve this goal and can be used for autorecovery.

Reply:
I save the `group` information in the custom metadata.

Currently, the `replaceBookie` will get all the bookies that are not in the primary group and secondary group and add them to the blacklist.
Do you mean I should not fill the primary and secondary isolation groups with the ensemble placement policy config which read from the custom metadata rather that I should directly use the placement policy config to get a blacklist?

Reply:
@sijie Please take a look, thanks.

Reply:
@zymap a local var should be enough